Analysis/bin/ShuffleMergeSpectral.cxx

```diff
@@ -0,0 +1,681 @@
+/*! Sheffle and merge datasets using Histograms created with CreateSpectralHists.cxx
+*/
+
+#include <fstream>
+#include <random>
+#include <string>
+#include <iostream>
+#include <boost/algorithm/string.hpp>
+#include <boost/regex.hpp>
+#include <boost/foreach.hpp>
+
+#include "TauMLTools/Core/interface/program_main.h"
+#include "TauMLTools/Core/interface/RootExt.h"
+#include "TauMLTools/Core/interface/PropertyConfigReader.h"
+#include "TauMLTools/Core/interface/ProgressReporter.h"
+#include "TauMLTools/Analysis/interface/TauTuple.h"
+
+namespace analysis {
+
+enum class MergeMode { MergeAll = 1, MergePerEntry = 2 };
+ENUM_NAMES(MergeMode) = {
+    { MergeMode::MergeAll, "MergeAll" },
+    { MergeMode::MergePerEntry, "MergePerEntry" }
+};
+} // namespace analysis
+
+struct Arguments {
+    run::Argument<std::string> cfg{"cfg", "configuration file with the list of input sources"};
+    run::Argument<std::string> input{"input", "input path with tuples for all the samples"};
+    run::Argument<std::string> output{"output", "output, depending on the merging mode: MergeAll - file,"
+                                                " MergePerEntry - directory."};
+    run::Argument<std::string> pt_bins{"pt-bins", "pt bins"};
+    run::Argument<std::string> eta_bins{"eta-bins", "eta bins"};
+    run::Argument<analysis::MergeMode> mode{"mode", "merging mode: MergeAll or MergePerEntry"};
+    run::Argument<bool> calc_weights{"calc-weights", "calculate training weights"};
+    // run::Argument<bool> ensure_uniformity{"ensure-uniformity", "stop the merge if at least one of the bins is empty"};
+    run::Argument<size_t> max_bin_occupancy{"max-bin-occupancy", "maximal occupancy of a bin",
+                                            std::numeric_limits<size_t>::max()};
+    run::Argument<unsigned> n_threads{"n-threads", "number of threads", 1};
+    run::Argument<unsigned> seed{"seed", "random seed to initialize the generator used for sampling", 1234567};
+    run::Argument<std::string> disabled_branches{"disabled-branches",
+                                                 "list of branches to disabled in the input tuples", ""};
+    run::Argument<std::string> path_spectrum{"input_spec", "input path with spectrums for all the samples"};
+    run::Argument<std::string> tau_ratio{"tau_ratio", "ratio of tau types in the final spectrum"};
+};
+
+namespace {
+struct SourceDesc {
+    using Tau = tau_tuple::Tau;
+    using TauTuple = tau_tuple::TauTuple;
+    using SampleType = analysis::SampleType;
+
+    SourceDesc(const std::string& _name,const std::string&  _group_name, const std::vector<std::string>& _file_names,
+               double _weight, const std::set<std::string>& _disabled_branches) :
+        name(_name), group_name(_group_name), file_names(_file_names), disabled_branches(_disabled_branches), weight(_weight),
+        current_n_processed(0), files_n_total(_file_names.size()), entries_file(666), total_n_processed(0)
```

Comment:
To make code parallelizable, an entry range should be passed. Since each file has a different size,  the range should be defined in relative values.